### ISSUE TITLE
False positive: EmptyLineSeparator report wrong violation in the case of anonymous class

### ISSUE THREAD

**Issue by Praveen7294** [checkstyle/checkstyle#18858]
detected at https://github.com/checkstyle/checkstyle/pull/18545#issuecomment-3822474864

I have read check documentation: https://checkstyle.org/checks/whitespace/emptylineseparator.html#EmptyLineSeparator
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

`EmptyLineSeparatorCheck` reports wrong violation or two violation pointing to same line in the case of anonymous class.

#### Config:
```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
   "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="EmptyLineSeparator">
      <property name="tokens"
               value="PACKAGE_DEF, IMPORT, STATIC_IMPORT, CLASS_DEF, INTERFACE_DEF, ENUM_DEF,
                    STATIC_INIT, INSTANCE_INIT, METHOD_DEF, CTOR_DEF, VARIABLE_DEF, RECORD_DEF,
                    COMPACT_CTOR_DEF"/>
      <property name="allowNoEmptyLineBetweenFields" value="true"/>
      <property name="allowMultipleEmptyLines" value="false"/>
      <property name="allowMultipleEmptyLinesInsideClassMembers" value="false"/>
    </module>
  </module>
</module>
```
Taking three input example with anonymous class for better understanding:
---
### Case 1:
Anonymous class with mutiple chained method call
#### Test.java
```
import java.util.Stack;

public class Test {

    public static void foo1() {
        Stack<Integer> stack = new Stack<>();
        stack.push(5);
 	    new EarlyTerminatingScanner() {
             @Override
             protected void visit(Object role, Object element) {
                 // violation 'There is more than 1 empty line after this line.'
             }       // line  12


 	    }
 	    .setEnabled(true)
 	    .scan("ROLE", new Object());     // line 17
    }

    static abstract class EarlyTerminatingScanner {

        private boolean enabled;

        public EarlyTerminatingScanner setEnabled(boolean value) {
            enabled = value;
            return this;
        }

        public void scan(Object role, Object element) {
            if (enabled) {
                visit(role, element);
            }
        }

        protected abstract void visit(Object role, Object element);
    }
}
```
#### cli:
```
$ java -jar checkstyle-13.1.0-all.jar -c config_check.xml Test.java
Starting audit...
[ERROR] F:\GitHub\headhtmltagname\Test.java:12:14: '}' has more than 1 empty lines after. [EmptyLineSeparator]
[ERROR] F:\GitHub\headhtmltagname\Test.java:17:13: There is more than 1 empty line after this line. [EmptyLineSeparator]
Audit done.
Checkstyle ends with 2 errors.
```

### Case 2:
Anonymous class with single chained method call

#### Test.java
```
import java.util.Stack;

public class Test {

    public static void foo1() {
        Stack<Integer> stack = new Stack<>();
        stack.push(5);
 	    new EarlyTerminatingScanner() {
             @Override
             protected void visit(Object role, Object element) {
                 // violation 'There is more than 1 empty line after this line.'
             }      // line 12


 	    }
 	    .setEnabled(true);
    }

    static abstract class EarlyTerminatingScanner {

        private boolean enabled;

        public EarlyTerminatingScanner setEnabled(boolean value) {
            enabled = value;
            return this;
        }

        protected abstract void visit(Object role, Object element);
    }
}
```

#### cli:
```
$ java -jar checkstyle-13.1.0-all.jar -c config_check.xml Test.java
Starting audit...
[ERROR] F:\GitHub\headhtmltagname\Test.java:12:14: '}' has more than 1 empty lines after. [EmptyLineSeparator]
[ERROR] F:\GitHub\headhtmltagname\Test.java:12:14: There is more than 1 empty line after this line. [EmptyLineSeparator]
Audit done.
Checkstyle ends with 2 errors.
```

### Case 3:
Only anonymous class, no chained method calls

#### Test.java
```
import java.util.Stack;

public class Test {

    public static void foo1() {
        Stack<Integer> stack = new Stack<>();
        stack.push(5);
 	    new EarlyTerminatingScanner() {
             @Override
             protected void visit(Object role, Object element) {
                 // violation 'There is more than 1 empty line after this line.'
             }      // line 12


 	    };
    }

    static abstract class EarlyTerminatingScanner {

        protected abstract void visit(Object role, Object element);
    }
}
```
#### cli:
```
$ java -jar checkstyle-13.1.0-all.jar -c config_check.xml Test.java
Starting audit...
[ERROR] F:\GitHub\headhtmltagname\Test.java:12:14: '}' has more than 1 empty lines after. [EmptyLineSeparator]
[ERROR] F:\GitHub\headhtmltagname\Test.java:12:14: There is more than 1 empty line after this line. [EmptyLineSeparator]
Audit done.
Checkstyle ends with 2 errors.
```
---

`EmptyLineSeparatorCheck` report wrong violation on `case 1` and two violation pointing to the same line in `case 2` & `case 3`. In my opinion, `EmptyLineSeparatorCheck` with `<property name="allowMultipleEmptyLinesInsideClassMembers" value="false"/>` have no handling case of anonymous class.

**Comment by Gaoyan1999** [comment]
@Praveen7294 
Will you fix this issue on https://github.com/checkstyle/checkstyle/pull/18545 as well?  If not, I'm happy to be on it.

**Comment by Praveen7294** [comment]
> Will you fix this issue on https://github.com/checkstyle/checkstyle/pull/18545 as well? If not, I'm happy to be on it.

let first merge #18545 to avoid merge conflicts then you can work.

**Comment by Anushreebasics** [comment]
@Praveen7294 are u working on this?

**Comment by Praveen7294** [comment]
@Anushreebasics Yes you can, but my PR related to similar issue not yet merged may be you will get merge conflict in your PR.

**Comment by achyuth8055** [comment]
Hello @Praveen7294,

I noticed there was a discussion earlier about working on this issue. I wanted to check if it is still being worked on.
If it is currently unassigned or no one is actively working on it, I would be happy to take a look and work on a fix.
Please let me know if that would be okay.

Thank you!

**Comment by Praveen7294** [comment]
> I noticed there was a discussion earlier about working on this issue. I wanted to check if it is still being worked on.
If it is currently unassigned or no one is actively working on it, I would be happy to take a look and work on a fix.
Please let me know if that would be okay.

Yeah you can work but you are new to checkstyle. First start with good first issues it will be helpful to understand the development of core.

Please [read](https://github.com/checkstyle/checkstyle/blob/master/.github/CONTRIBUTING.md#getting-started) and watch videos at [Starting_Development](https://checkstyle.org/beginning_development.html#Starting_Development).

**Comment by achyuth8055** [comment]
@Praveen7294 Thanks for the guidance! I'll start with a few good-first issues to get familiar with the codebase and then come back to this issue if it's still open.

**Comment by pushnanashi2** [comment]
I am on it.
